### PR TITLE
Refactor GitHubAppTokenProvider

### PR DESCRIPTION
**Pull request by epananth**
Issue -> https://github.com/dotnet/core-eng/issues/9582

This change will make it easier to mock GithubAppTokenProvider in future

### CODE REVIEW

File: src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.DotNet.GitHub.Authentication
+{
+    public interface IGitHubAppTokenProvider 
+    {
+        string GetAppToken();
+        string GetAppTokenFromEnvironmentVariableBase64(int gitHubAppId, string environmentVariableName);
```
Comment (ChadNedzlek) [inline src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:10]: This should probably just be a different implementation of the provider.  I wouldn't personally add this to the interface.  The "DarcBot" should just choose a different IGutHubAppTokenProvider that does the right thing.

Comment (ChadNedzlek) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:10]: I have lots of comments about the implementation of that method, but for now, just leaving it off the interface is the best option. :-)

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:10]: I'm confused this was the original comment -> https://github.com/dotnet/arcade-services/pull/1084#discussion_r405812677, I extracted this to an interface. Now I feel the comments are conflicting.

Comment (ChadNedzlek) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:10]: Extract the method you want to the new interface. I didn't realize this other weird method exists.  It's not a good abstraction, and you don't need it for testing.  You don't need every method to be in the interface.

Honestly, I'd rather we delete that other method and fix it's implementation to be better encapsulated, but that change doesn't impact the testing PR at all, so it's not really worth it for you to do right now.  Like it doesn't make sense for it to be told an environment variable key for it to pull it out... why didn't the caller just give us the answer instead of binding this method to the environment (especially since the first argument, gitHubAppid, is ALSO pulled from the environment... they pulled one of them but not the other)...  But all of that is sort of silly, because the implementation of this method doesn't use any instance data, it should just be a static method somewhere.   But really really, there should be a "EnvironmentGitHubAppTokenProvider" that implements the new interface, and just implements the "GetAppToken" appropriately.  The whole point of the original class was that the caller didn't need all the gory details... the added method takes all the gory details and destroys the whole point of the class.

TL;DR: Remove the GetAppTokenFromEnvironmentVariableBase64 method, leave the GetAppToken method.

Comment (riarenas) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:10]: The important method in the interface is GetAppToken, right? If I understand what @ChadNedzlek means, I think he's right. The Environment variable method should probably not be in the interface, and there would be a `GithubAppTokenProviderFromEnvironment` implementation.

You can just leave it in the class

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs]: I am going to create another issue to do this cleanup .

File: src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs [hunk 2]
```diff
@@ -3,12 +3,13 @@
 // See the LICENSE file in the project root for more information.
 
 using GitHubJwt;
+using Microsoft.DotNet.GitHub.Authentication;
```
Comment (riarenas) [inline src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: Is this using necessary?

Comment (riarenas) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: Was it merged with the using statement? this is something that should probably be cleaned up as it seems weird to add a using for the same namespace that the class is in.

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: Sorry my email on the comment came later. :( I will clean it up.

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: Actually those are 2 different ones 

1) Microsoft.DotNet.GitHub.Authentication;
2) Microsoft.Dotnet.GitHub.Authentication

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: There is this this weird naming we have all over.
Microsoft.DotNet.GitHub.Authentication has all our classes .  Microsoft.Dotnet.GitHub.Authentication is Github lib namespace. So it looks like they are the same but they actually are not.  :)

Comment (riarenas) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: Is that just a typo in this file?

Comment (riarenas) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: Because this class is ours and it's in the Dotnet namespace

Comment (riarenas) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: Ooh, maybe the interface is in the wrong namespace then? and it should be using the lowercase one?

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/GitHubAppTokenProvider.cs:6]: I see some files have DotNet other have Dotnet. Its all over in that folder. Need to fix that. I can create another PR to fix this.

File: src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.DotNet.GitHub.Authentication
```
Comment (riarenas) [inline src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:5]: Yeah I think this is the thing. This interface is introducing a new namespace instead of using the Dotnet with lowercase n that the classes are in.

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:5]: there are other files with the same prob -> 
image

but this is the folder structure 

image

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:5]: I will change all of them to use Microsoft.DotNet.GitHub.Authentication in that folder.

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:5]: ah you mentioned the same file :)

Comment (riarenas) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:5]: We're inconsistent in naming? say that ain't so! Thanks for taking a look haha.

Comment (epananth) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:5]: created an issue -> https://github.com/dotnet/core-eng/issues/9625, so resolving this comment.

Comment (riarenas) [reply src/Shared/Microsoft.DotNet.GitHub.Authentication/IGitHubAppTokenProvider.cs:5]: Thanks!
